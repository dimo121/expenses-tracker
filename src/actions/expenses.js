import uuid from 'uuid';
import database from '../firebase/firebase';

// Including a database design the action generators will return functions instead of
// objects. Redux does not have usually allow this functionality so we use redux thunk middleware
// to add support for redux to execute the function e.g. firebase push to add object.

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    database.ref(`expenses/${id}`).remove()
    .then(() => {
      dispatch(removeExpense({ id }));
    });
  }
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});


export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    database.ref(`expenses/${id}`).update(updates)
    .then(() => {
      dispatch(editExpense(id,updates));
    });
  }
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

//SET_EXPENSES

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});


export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

      dispatch(setExpenses(expenses));

    });
  };
};

