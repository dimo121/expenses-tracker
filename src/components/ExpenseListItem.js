import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';


//need an expense list item header that takes in filtered expenses and 
//sends back the number of epxenses being shown and their total sum.
//need getExpensesTotal function from expenses-total.js
//should return 0 if no expenses
//should correctly add up a single expense
//should correctly add up multiple expenses
//the header will sit in between the navbar and the expense form
//Need to test these functionalities

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <Link className='list-item' to={`/edit/${id}`}>
    <div>
      <h3 className='list-item__title'>{description}</h3>
      <span className='list-item__sub-title'>{moment(createdAt).format('Do MMMM, YYYY')}</span>
    </div>
    <h3 className='list-item__data'>{numeral(amount / 100).format('$0,0.00')}</h3>
  </Link>
);

export default ExpenseListItem;
