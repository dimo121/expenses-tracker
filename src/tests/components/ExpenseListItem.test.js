import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixutres/expenses';

test('should test expenseListItem component', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]} />)
    expect(wrapper).toMatchSnapshot();
});


