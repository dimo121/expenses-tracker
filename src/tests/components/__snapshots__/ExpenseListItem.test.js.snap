// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test expenseListItem component 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/89374edd-2b99-4974-a860-9215c211307f"
  >
    <h3>
      Water bill
    </h3>
  </Link>
  <p>
    $45.00
    -
    1st January, 1970
  </p>
</div>
`;
